Performance tuning

leanil opened the pull request:
Fixed a bug in server-side cache updating.
Parallelized the fetching of reports over projects.

paldys commented on server/src/cache.js:
```diff
@@ -6,7 +6,7 @@ export function create(load) {
-  if (cache.storage[key] === undefined) await update
+  const update = () => cache.load(key, args).then((v) => {cache.storage[key] = v})
+  if (cache.storage[key] === undefined) await update()
   return cache.storage[key]
```
Do you understand why was it this way? How did this work?

What you do now, is check the cache, if it exists return its content, otherwise load it (from Slack). What you miss here, is the fact that the data on Slack might have changed in the meantime (e.g. replies for a thread). This is why it worked as it worked: you fire the promise which will update the cache to the current state, but don't wait for it and just return the cached content if there is any.

paldys reviewed (commented):
I think we can still do better than this, what's the point of loading everything one-by-one?

There is a lot of duplicated content (emojis, channels, users..). Can you investigate the option to gather everything on the server?

leanil committed:
Load all reports at the first request to the page

Change server api: load all the reports for all
the tags in one request.
Warm up caches at startup.

Resolves #34

leanil committed:
Add 5 mins Time To Live to server slack cache

Store the time of the last fetch for each key in the cache.
Don't refresh the cache on request if the last fetch was
within a given time frame (currently 5 minutes). This is
necessary because the slack api is rate limited. See more:
https://api.slack.com/docs/rate-limits

leanil committed:
Reload all reports in every 5 minutes

Use the render prop of Route instead of component, not to re-create
the Content component at every render.
https://reacttraining.com/react-router/web/api/Route/render-func
This makes it possible to use Content's shouldComponentUpdate to
prevent re-render (and scrolling to the end) if there are no changes.

leanil commented:
Update report loading and cache logic.

paldys commented on client/src/App.js:
```diff
@@ -10,15 +10,32 @@ import './App.scss'
+      ({tag, reports}) =>
+        {reportsByTag[tag] = reports.map((ts) => reportByTs[ts])}
+    )
+    reportData.reportsByTag = reportsByTag
```
I think doing this will be easier on the server, I will explain it in the other comment.

paldys commented on server/src/db.js:
```diff
@@ -5,19 +5,18 @@ import {Client} from 'pg'
-     WHERE tag.tag = $1
-     ORDER BY ts ASC`,
-    [tag],
+     FROM report
```
Do I understand correctly, that there are reports in the database that were never read before? Can you check and confirm this to me, please? I am might have misunderstood it, but it looks to me, that every comment is saved in the database, regardless if it has a tag or not.

If that's the case, we don't need to load all the reports here. It's unnecessary. You would need to change the original query to something like:
```
SELECT ts, tag.tag, "user", message, channel, response_to
     FROM tag LEFT JOIN report ON tag.report=report.ts
     ORDER BY ts ASC
```

So, only the tagged reports are loaded.

leanil replied:
Yes, untagged messages are also stored in the db.

paldys commented on server/src/index.js:
```diff
@@ -83,8 +83,8 @@ function normalizeEmoji(emoji) {
-app.get('/api/reports-by-tags/:tag', authorize, async (req, res) => {
-  let reports = await loadReportsByTag(req.params.tag)
+async function loadReportData() {
+  let reports = await loadReports()
```
There might be some overlap between reports, but I am assuming it will be minimal and the overlap itself is handled quite well because of our cache (their replies are not loaded two times).

paldys commented on server/src/index.js:
```diff
@@ -103,18 +103,23 @@ app.get('/api/reports-by-tags/:tag', authorize, async (req, res) => {
 
 
-  res.json({
+  return {
```
When you return the `reports`, let's do some nice [groupBy](https://lodash.com/docs/4.17.15#groupBy) `tag` on them before sending them to the UI. (this should replace all the computing on the client you had)

leanil commented on client/src/components/content/Content.js:
```diff
@@ -1,37 +1,15 @@
-    const response = await getReportsByTag(this.props.tag)
-    this.setState({...response, loading: false})
-  }
-
```
I think this loading state doesn't make much sense in this component now, that this component is only rendered once everything has been downloaded.

paldys commented on client/src/App.js:
```diff
@@ -12,13 +12,13 @@ class App extends Component {
     const tags = await getTags()
-    this.setState({tags})
+    const reportData = await getReportData()
+    this.setState({tags, reportData});
```
(nit) You left a semicolon on the end.

paldys commented on client/src/App.js:
```diff
@@ -30,7 +30,7 @@ class App extends Component {
               const tag = decodeURI(props.match.params.tag)
               return tags.some((t) => t.tag === tag)
-                ? (<Content key={tag} tag={tag} />)
+                ? (<Content key={tag} tag={tag} reportData={reportData}/>)
```
We cand send only the reports we need because we know the tag.

paldys commented on client/src/components/content/Content.js:
```diff
@@ -1,37 +1,15 @@
-    const {tag} = this.props
-    const {emoji, reports, users, channels, loading} = this.state
-
+    const {tag, reportData: {reports, emoji, users, channels}} = this.props
```
This component will be re-rendered and `MessageList` will scroll to the end every 5 minutes, even if there wasn't a change.

paldys commented on client/src/components/content/Content.js:
```diff
@@ -1,37 +1,21 @@
-    const response = await getReportsByTag(this.props.tag)
-    this.setState({...response, loading: false})
+  shouldComponentUpdate(nextProps) {
+    return !_.isEqual(this.props.reportData.reports, nextProps.reportData.reports)
```
Can there be a possibility that reports stay the same but tag changes? :slightly_smiling_face: ..I think it would be more correct to add check for `tag` too.

paldys commented on client/src/App.js:
```diff
@@ -27,10 +36,11 @@ class App extends Component {
+              const reportDataForTag = {...reportData, reports: reportData.reports[tag]}
               return tags.some((t) => t.tag === tag)
-                ? (<Content key={tag} tag={tag} />)
+                ? (<Content key={tag} tag={tag} reportData={reportDataForTag} />)
```
Why not send deconstructed variables right away (reports, users... )?

paldys commented:
Please, investigate the usage of comments stored in the database. 

While this solution is technically good and it seems to work and the caching introduced here helps a lot. However, we can still hit the rate limits of the slack web API easily and it could cause more problems in the future. The bottleneck seems to be the API for [`conversations.replies`](https://api.slack.com/methods/conversations.replies) which is in [Tier 3](https://api.slack.com/docs/rate-limits#tier_t3). We would need circa 500 calls to this, but it's limited to 50+ per minute.

It seems like all the information, we would need from `conversations.replies` is already stored in the database. Let's investigate if this is true or if there are any disadvantages to this solution.

leanil committed:
Load replies from DB instead of slack API

The conversations.replies API call caused a bottleneck, because
it's in Tier 3 (50 calls / minute) and we use it on every report.
As the client needs only such info about replies that we already
have in the DB, we can completely skip the API call.

paldys commented on server/src/index.js:
```diff
@@ -83,38 +83,41 @@ function normalizeEmoji(emoji) {
-      await cache.get(
-        replies, r.response_to || r.ts, r.channel
-      )).messages,
+    replies: [...await loadReplies(r.response_to || r.ts)],
```
This point is my only concern about this solution. We are hitting the database nearly 500 times, which is not nice. 

Do you think we can optimize this? 

A really easy change would be for `loadReplies` to load everything (sorted by `ts` as it is now) and go:

```
  const repliesByThread = _.groupBy(await loadReplies(), (r) => r.response_to || r.ts)
  let reports = (await loadReports()).map((r) => ({
    ...r,
    replies: repliesByThread[r.response_to || r.ts]
  }))
```

paldys reviewed (commented):
One single concern I had, what do you think about it?

paldys reviewed (approved):
:+1:
